In [70]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [71]:
df = pd.read_csv('/content/train_data.csv')
df_test = pd.read_csv('/content/test_data.csv')
df.head()


x1  x2  x3  x4          x5        x6        x7          x8        x9  \
0   0   0   0   1 -323.106605  2.694366 -1.987520  261.092784  0.013704   
1   0   0   0   1 -376.084691  0.969696 -6.933765  355.311648  0.030292   
2   0   0   0   0   91.955425  2.621643 -2.581162   51.357206  0.036668   
3   0   0   0   1 -391.814586  1.866914 -2.510799  382.900317  0.007947   
4   0   0   0   0 -363.823732  2.951346 -3.726368  330.527539  0.010074   

        x10           x11       x12       x13  x14       x15       x16  \
0  0.000100 -8.900000e-06 -0.806753 -0.805752  401  0.043217  0.001868   
1 -0.000153  8.830000e-07 -0.940358 -0.937867  401  0.080432  0.006469   
2 -0.000104 -9.220000e-05  0.240181  0.229315  401  0.274750  0.075487   
3 -0.000028  2.990000e-05 -0.978424 -0.977094  401  0.012307  0.000151   
4 -0.000004 -2.060000e-05 -0.915454 -0.907291  401  0.032879  0.001081   

        x17       x18        x19  x20  x21  x22  x23       x24       x25  \
0  0.012497 -0.463844   5.481409   67   78  193  208  0.551122  0.548628   
1 -0.840971  5.311628  12.116713   45   65  194  207  0.384040  0.381546   
2 -0.062232 -0.579688  14.667095   52   48  204  197  0.411471  0.408978   
3  0.227388 -0.048079   3.178819   36   35  185  216  0.027431  0.024938   
4  0.746032  0.262103   4.029558   94  106  176  225  0.082294  0.079800   

        x26       x27       x28       x29       x30       x31       x32  \
0  0.127182  0.124688  0.002506  0.007481 -0.746494 -2.239481  0.995012   
1  0.376559  0.374065  0.002500  0.004988 -0.883335 -1.766671  0.997506   
2  0.324190  0.321696  0.000000  0.000000  0.000000  0.000000  1.000000   
3  0.379052  0.376559  0.012626  0.024938 -4.876954 -9.753908  0.987531   
4  0.503741  0.501247  0.000000  0.000000  0.000000  0.000000  1.000000   

        x33       x34       x35       x36       x37       x38  x39  x40  x41  \
0  2.172739 -0.689267 -0.906939  0.000347  0.000291  0.000559    1    1    1   
1  1.803625 -0.642837 -1.362523  0.005266  0.004333 -0.003138    1    1    1   
2  2.121199  0.818939 -0.345764 -0.001580 -0.003453 -0.005586    1    1    1   
3  2.185079 -0.936042 -1.012427 -0.000084 -0.000405 -0.000674    1    1    1   
4  2.063019 -0.774391 -0.971482  0.000072  0.000002  0.000007    1    1    1   

   x42  x43  x44  x45  x46  x47  x48       x49       x50       x51       x52  \
0    1    0    0    0    0    0    0 -0.863371 -0.844213 -0.825756 -0.815687   
1    0    0    0    0    0    0    0 -1.010609 -0.973787 -0.961984 -0.951379   
2    1    1    0    0    0    0    0 -0.164767 -0.032252  0.100638  0.198381   
3    1    0    0    0    0    0    0 -0.992626 -0.986964 -0.983844 -0.981017   
4    1    0    0    0    0    0    0 -0.943286 -0.935992 -0.930382 -0.924657   

        x53       x54       x55       x56  x57       x58       x59       x60  \
0 -0.797841 -0.783021 -0.766101 -0.746494    1  0.897371  0.839176  0.811325   
1 -0.923490 -0.901670 -0.886030 -0.861417    1  0.769634  0.307352 -0.039063   
2  0.280628  0.349402  0.476402  0.610916    1  0.983643  0.956236  0.930309   
3 -0.974313 -0.970775 -0.966204 -0.960642    1  0.639823  0.457482  0.394356   
4 -0.902470 -0.888912 -0.878450 -0.863476    1  0.914871  0.839502  0.779439   

        x61       x62       x63       x64       x65       x66  x67  x68  x69  \
0  0.774266  0.730998  0.704487  0.676231  0.649431  0.648554   28   33  109   
1 -0.120251 -0.014816  0.111915  0.139059  0.059254 -0.061575   28   33   80   
2  0.902098  0.869844  0.835157  0.799799  0.762408  0.723252   30   31   82   
3  0.415160  0.447908  0.549856  0.525067  0.484631  0.388960   36   40  109   
4  0.715020  0.649401  0.600535  0.554517  0.520204  0.502750   35   41  110   

   x70  x71  x72  x73  x74       x75       x76       x77       x78       x79  \
0   42   30    1    1    1  2.079063  0.009790  0.012405  0.012023  0.013513   
1   47   31    1    1    1  1.407760  0.047294  0.030238  0.024851  0.025847   
2   24   13    1    1    1  2.19374

In [72]:
df_test.shape, df.shape

((1732, 669), (4584, 670))

In [73]:
# check for missing data
for col in df.columns:
    if df[col].isna().any():
        print(col)

print('No Missing Data in Training Set')


No Missing Data in Training Set


In [74]:
# find the value counts for the target variable, phase, and state
print(df['output'].value_counts(normalize=True) * 100)


output
1    84.489529
0    15.510471
Name: proportion, dtype: float64


In [75]:
# Create DataFrames for subject and state counts
train_subject = df['subject'].value_counts().reset_index()
train_state = df['state'].value_counts().reset_index()
test_subject = df_test['subject'].value_counts().reset_index()
test_state = df_test['state'].value_counts().reset_index()

# Rename columns
train_subject.columns = ['Subject', 'Train Count']
train_state.columns = ['State', 'Train Count']
test_subject.columns = ['Subject', 'Test Count']
test_state.columns = ['State', 'Test Count']

# Merge DataFrames
subject_table = pd.merge(train_subject, test_subject, on='Subject', how='outer').fillna(0)
state_table = pd.merge(train_state, test_state, on='State', how='outer').fillna(0)

# Convert counts to integers
subject_table[['Train Count', 'Test Count']] = subject_table[['Train Count', 'Test Count']].astype(int)
state_table[['Train Count', 'Test Count']] = state_table[['Train Count', 'Test Count']].astype(int)

print("Subject Counts:")
print(subject_table.to_string(index=False))
print("\nState Counts:")
print(state_table.to_string(index=False))

Subject Counts:
Subject  Train Count  Test Count
      A          380          59
      B          513          78
      C          361          71
      D          420          83
      E            0         312
      F          330          60
      G          518          88
      H          181          33
      I          663         124
      J            0         629
      K          468          69
      L          444          71
      M          306          55

State Counts:
State  Train Count  Test Count
    A          398         112
    B         1502         525
    C         2106         841
    D          576         253
    E            2           1


#### Quick Observations
1. The dataset has 4584 rows and 670 columns.
2. The target variable is 'output' which has two classes: 0 and 1.
3. The dataset is imbalanced with respect to the target variable.
4. The dataset has no missing values.
5. The state variable seems a somewhat imbalanced ()
6. Disproportionate state E.

In [76]:
print(f"In the train set: {sorted(df['subject'].unique())}")
print(f"In the test set: {sorted(df_test['subject'].unique())}")


In the train set: ['A', 'B', 'C', 'D', 'F', 'G', 'H', 'I', 'K', 'L', 'M']
In the test set: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M']


There are some subjects in the test set that are not in the train set. We will be mapping unseen categories (e.g., subjects or states) to an 'unknown' category in both your training and test sets. This allows the model to handle unseen categories as a special case rather than failing outright.

In [77]:
print(f"In the train set: {sorted(df['state'].unique())}")
print(f"In the test set: {sorted(df_test['state'].unique())}")
# drop the subjects that are not in the train set

In the train set: ['A', 'B', 'C', 'D', 'E']
In the test set: ['A', 'B', 'C', 'D', 'E']


In [78]:
# Convert subjects and states to string type
df['subject'] = df['subject'].astype(str)
df_test['subject'] = df_test['subject'].astype(str)

# Handle unseen subjects in the test set by assigning them to 'unknown'
df_test['subject'] = df_test['subject'].apply(lambda x: x if x in df['subject'].unique() else 'unknown')

# Similarly, for states (if necessary)
df_test['state'] = df_test['state'].apply(lambda x: x if x in df['state'].unique() else 'unknown')


In [79]:
from sklearn.preprocessing import OneHotEncoder

# Define the categorical columns
categorical_features = ['state', 'subject']
ordinal_features = ['phase']
target_feature = ['output']

# numerical features
numerical_features = [col for col in df.columns if col not in categorical_features + ordinal_features + target_feature]

# Set up the encoder. OneHotEncoder using ha
encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False)

# Fit on the training data
X_train_encoded = encoder.fit_transform(df[categorical_features])

# Apply the same transformation to the test set
X_test_encoded = encoder.transform(df_test[categorical_features])


In [80]:
X_train_encoded.shape, X_test_encoded.shape

((4584, 16), (1732, 16))

In [81]:
# Get the new column names after encoding
encoded_columns = encoder.get_feature_names_out(categorical_features)

# Convert the encoded data back into a DataFrame to preserve column names
X_train_encoded_df = pd.DataFrame(X_train_encoded, columns=encoded_columns, index=df.index)
X_test_encoded_df = pd.DataFrame(X_test_encoded, columns=encoded_columns, index=df_test.index)

# concatenate these encoded features back with the original data (excluding the original categorical columns)
X_train_final = pd.concat([df.drop(columns=categorical_features), X_train_encoded_df], axis=1)
X_test_final = pd.concat([df_test.drop(columns=categorical_features), X_test_encoded_df], axis=1)


X_train_final.shape, X_test_final.shape

((4584, 684), (1732, 683))

In [82]:
pd.set_option('display.max_columns', None)
X_train_final.head()

x1  x2  x3  x4          x5        x6        x7          x8        x9  \
0   0   0   0   1 -323.106605  2.694366 -1.987520  261.092784  0.013704   
1   0   0   0   1 -376.084691  0.969696 -6.933765  355.311648  0.030292   
2   0   0   0   0   91.955425  2.621643 -2.581162   51.357206  0.036668   
3   0   0   0   1 -391.814586  1.866914 -2.510799  382.900317  0.007947   
4   0   0   0   0 -363.823732  2.951346 -3.726368  330.527539  0.010074   

        x10           x11       x12       x13  x14       x15       x16  \
0  0.000100 -8.900000e-06 -0.806753 -0.805752  401  0.043217  0.001868   
1 -0.000153  8.830000e-07 -0.940358 -0.937867  401  0.080432  0.006469   
2 -0.000104 -9.220000e-05  0.240181  0.229315  401  0.274750  0.075487   
3 -0.000028  2.990000e-05 -0.978424 -0.977094  401  0.012307  0.000151   
4 -0.000004 -2.060000e-05 -0.915454 -0.907291  401  0.032879  0.001081   

        x17       x18        x19  x20  x21  x22  x23       x24       x25  \
0  0.012497 -0.463844   5.481409   67   78  193  208  0.551122  0.548628   
1 -0.840971  5.311628  12.116713   45   65  194  207  0.384040  0.381546   
2 -0.062232 -0.579688  14.667095   52   48  204  197  0.411471  0.408978   
3  0.227388 -0.048079   3.178819   36   35  185  216  0.027431  0.024938   
4  0.746032  0.262103   4.029558   94  106  176  225  0.082294  0.079800   

        x26       x27       x28       x29       x30       x31       x32  \
0  0.127182  0.124688  0.002506  0.007481 -0.746494 -2.239481  0.995012   
1  0.376559  0.374065  0.002500  0.004988 -0.883335 -1.766671  0.997506   
2  0.324190  0.321696  0.000000  0.000000  0.000000  0.000000  1.000000   
3  0.379052  0.376559  0.012626  0.024938 -4.876954 -9.753908  0.987531   
4  0.503741  0.501247  0.000000  0.000000  0.000000  0.000000  1.000000   

        x33       x34       x35       x36       x37       x38  x39  x40  x41  \
0  2.172739 -0.689267 -0.906939  0.000347  0.000291  0.000559    1    1    1   
1  1.803625 -0.642837 -1.362523  0.005266  0.004333 -0.003138    1    1    1   
2  2.121199  0.818939 -0.345764 -0.001580 -0.003453 -0.005586    1    1    1   
3  2.185079 -0.936042 -1.012427 -0.000084 -0.000405 -0.000674    1    1    1   
4  2.063019 -0.774391 -0.971482  0.000072  0.000002  0.000007    1    1    1   

   x42  x43  x44  x45  x46  x47  x48       x49       x50       x51       x52  \
0    1    0    0    0    0    0    0 -0.863371 -0.844213 -0.825756 -0.815687   
1    0    0    0    0    0    0    0 -1.010609 -0.973787 -0.961984 -0.951379   
2    1    1    0    0    0    0    0 -0.164767 -0.032252  0.100638  0.198381   
3    1    0    0    0    0    0    0 -0.992626 -0.986964 -0.983844 -0.981017   
4    1    0    0    0    0    0    0 -0.943286 -0.935992 -0.930382 -0.924657   

        x53       x54       x55       x56  x57       x58       x59       x60  \
0 -0.797841 -0.783021 -0.766101 -0.746494    1  0.897371  0.839176  0.811325   
1 -0.923490 -0.901670 -0.886030 -0.861417    1  0.769634  0.307352 -0.039063   
2  0.280628  0.349402  0.476402  0.610916    1  0.983643  0.956236  0.930309   
3 -0.974313 -0.970775 -0.966204 -0.960642    1  0.639823  0.457482  0.394356   
4 -0.902470 -0.888912 -0.878450 -0.863476    1  0.914871  0.839502  0.779439   

        x61       x62       x63       x64       x65       x66  x67  x68  x69  \
0  0.774266  0.730998  0.704487  0.676231  0.649431  0.648554   28   33  109   
1 -0.120251 -0.014816  0.111915  0.139059  0.059254 -0.061575   28   33   80   
2  0.902098  0.869844  0.835157  0.799799  0.762408  0.723252   30   31   82   
3  0.415160  0.447908  0.549856  0.525067  0.484631  0.388960   36   40  109   
4  0.715020  0.649401  0.600535  0.554517  0.520204  0.502750   35   41  110   

   x70  x71  x72  x73  x74       x75       x76       x77       x78       x79  \
0   42   30    1    1    1  2.079063  0.009790  0.012405  0.012023  0.013513   
1   47   31    1    1    1  1.407760  0.047294  0.030238  0.024851  0.025847   
2   24   13    1    1    1  2.19374

In [83]:
X_test_final.head()

x1  x2  x3  x4          x5        x6         x7          x8        x9  \
0   0   0   0   1    4.665949  0.083412 -17.397081    0.067865  0.005895   
1   0   0   0   0 -398.299487  0.784032 -10.785768  403.784446  0.070405   
2   0   0   0   1 -380.902140  0.671324  -9.265829  361.841929  0.006937   
3   0   0   0   1 -342.784124  0.469414  -8.044287  293.043601  0.007217   
4   0   0   0   1    6.248502  0.252336  -5.398056    0.110097  0.005573   

        x10           x11       x12       x13  x14       x15       x16  \
0  0.000012 -2.090000e-05  0.011444  0.011636  401  0.005818  0.000034   
1  0.000180 -5.890000e-05 -0.977888 -0.993266  401  0.142714  0.020367   
2  0.000025 -5.380000e-06 -0.949965 -0.949881  401  0.008701  0.000076   
3 -0.000021  9.040000e-06 -0.854002 -0.854823  401  0.007698  0.000059   
4  0.000010  5.300000e-07  0.015457  0.015582  401  0.005635  0.000032   

        x17       x18        x19  x20  x21  x22  x23       x24       x25  \
0  0.034087  0.327009   2.358097    9   13  197  204  0.324190  0.321696   
1 -0.319843 -0.281809  28.161913   11   11  215  186  0.927681  0.925187   
2  0.337449  0.615012   2.774954   12   12  198  203  0.099751  0.097257   
3 -0.257195 -0.010352   2.886832   10   15  214  187  0.291771  0.289277   
4  0.047862 -0.299650   2.229109    8    8  197  204  0.835411  0.832918   

        x26       x27       x28       x29        x30        x31       x32  \
0  0.553616  0.551122  0.012723  0.032419   0.068020   0.175635  0.980050   
1  0.663342  0.660848  0.000000  0.000000   0.000000   0.000000  1.000000   
2  0.623441  0.620948  0.005025  0.012469  -1.897143  -4.749610  0.992519   
3  0.289277  0.286783  0.058511  0.117207 -18.785227 -40.126251  0.937656   
4  0.688279  0.685786  0.007595  0.022444   0.036548   0.106599  0.985037   

        x33       x34       x35           x36           x37           x38  \
0  2.154294  0.030290 -0.007108 -1.860000e-07 -4.480000e-08  8.590000e-08   
1  2.204332 -0.645101 -1.454337  1.375799e-02  1.996644e-02  1.846862e-02   
2  2.163365 -0.914883 -0.975463  4.380000e-05  9.340000e-05  8.180000e-05   
3  2.177239 -0.831585 -0.875648 -5.750000e-05 -1.061480e-04 -6.650000e-05   
4  2.221545  0.031521 -0.001468 -2.930000e-07  1.530000e-07 -3.230000e-08   

   x39  x40  x41  x42  x43  x44  x45  x46  x47  x48       x49       x50  \
0    1    1    1    1    0    0    0    0    0    0  0.003968  0.007107   
1    1    1    1    1    0    0    0    0    0    0 -1.185799 -1.123612   
2    1    1    1    0    0    0    0    0    0    0 -0.960480 -0.957066   
3    1    1    1    1    0    0    0    0    0    0 -0.865476 -0.861306   
4    1    1    1    1    0    0    0    0    0    0  0.008355  0.010875   

        x51       x52       x53       x54       x55       x56  x57       x58  \
0  0.009134  0.010318  0.012917  0.014480  0.016416  0.019016    1  0.138492   
1 -1.065856 -1.013341 -0.948284 -0.912662 -0.866746 -0.805297    1  0.785036   
2 -0.954753 -0.952626 -0.947893 -0.945974 -0.943289 -0.939138    1  0.467677   
3 -0.858338 -0.856174 -0.852460 -0.850653 -0.848182 -0.845795    1  0.235318   
4  0.012501  0.014046  0.017007  0.018515  0.020571  0.023094    1  0.232341   

        x59       x60       x61       x62       x63       x64       x65  \
0 -0.052568 -0.056833 -0.067865 -0.009167  0.059678  0.014958 -0.023078   
1  0.469576  0.188443 -0.002875 -0.161668 -0.283899 -0.397528 -0.471080   
2  0.298143  0.180085  0.063116  0.033894 -0.035738 -0.070120  0.010165   
3  0.094647  0.260005  0.073091  0.044456  0.098885  0.103982  0.099490   
4 -0.090812  0.021111  0.039492  0.055979 -0.000769  0.079791 -0.023135   

        x66  x67  x68  x69  x70  x71  x72  x73  x74       x75       x76  \
0  0.011155   40   37  121   56   37    1    1    1  1.851898  0.003125   
1 -0.492632   23   32   65   39   28    1    1    1  1.982182  0.072363   
2  0.036840   34   38  116   48   30    1    1    1  1.765693  0.003882   
3  0.076674   39   32  125   54   3

In [87]:
df_test.shape, df.shape

((1732, 669), (4584, 670))

### Question: Which Data is Sparse?

In [88]:
import pandas as pd

def compare_zero_proportions(train_df, test_df, threshold):
    """
    Compares the proportion of zeros between the train and test datasets
    and returns a list of columns where both datasets have sparse values
    above the given threshold.
    """
    train_zero_proportion = train_df.eq(0).sum() / train_df.shape[0]
    test_zero_proportion = test_df.eq(0).sum() / test_df.shape[0]

    # Find columns where both the train and test sets have proportions of zeros above the threshold
    sparse_columns = train_zero_proportion[train_zero_proportion > threshold].index
    sparse_in_test = test_zero_proportion[sparse_columns][test_zero_proportion[sparse_columns] > threshold].index

    return list(sparse_in_test)

In [92]:
# using a threshold of 0.99
sparse_columns_to_drop = compare_zero_proportions(X_train_final[numerical_features], df_test[numerical_features], threshold=0.99)

print("Columns to drop:", sparse_columns_to_drop)
print(f" We are dropping {len(sparse_columns_to_drop)} columns")

df_train_dropped = X_train_final.drop(columns=sparse_columns_to_drop, axis=1)
df_test_dropped = X_test_final.drop(columns=sparse_columns_to_drop, axis=1)

# print shapes before and after dropping
print(f"Shape of the training dataframe before dropping: {df.shape}")
print(f"Shape of the training dataframe after dropping: {df_train_dropped.shape}")
print(f"Shape of the testing dataframe before dropping: {df_test.shape}")
print(f"Shape of the testing dataframe after dropping: {df_test_dropped.shape}")

# check that the remaining columns are the same, except for the target variable
train_columns = set(df_train_dropped.columns) - {'output'}
test_columns = set(df_test_dropped.columns)

# Ensure that both training and test sets have the same columns
assert train_columns == test_columns, "Training and test columns do not match"

Columns to drop: ['x1', 'x2', 'x3', 'x46', 'x47', 'x48', 'x81', 'x86', 'x87', 'x91', 'x92', 'x93', 'x96', 'x97', 'x98', 'x99', 'x101', 'x102', 'x103', 'x104', 'x105', 'x113', 'y1', 'y2', 'y3', 'y46', 'y47', 'y48', 'y81', 'y86', 'y87', 'y91', 'y92', 'y93', 'y96', 'y97', 'y98', 'y99', 'y102', 'y103', 'y104', 'y105', 'y113', 'z1', 'z2', 'z3', 'z46', 'z47', 'z48', 'z81', 'z86', 'z87', 'z91', 'z92', 'z93', 'z96', 'z97', 'z98', 'z99', 'z101', 'z102', 'z103', 'z104', 'z105', 'z113']
 We are dropping 65 columns
Shape of the training dataframe before dropping: (4584, 670)
Shape of the training dataframe after dropping: (4584, 619)
Shape of the testing dataframe before dropping: (1732, 669)
Shape of the testing dataframe after dropping: (1732, 618)


In [93]:
df_train_dropped.head()

x4          x5        x6        x7          x8        x9       x10  \
0   1 -323.106605  2.694366 -1.987520  261.092784  0.013704  0.000100   
1   1 -376.084691  0.969696 -6.933765  355.311648  0.030292 -0.000153   
2   0   91.955425  2.621643 -2.581162   51.357206  0.036668 -0.000104   
3   1 -391.814586  1.866914 -2.510799  382.900317  0.007947 -0.000028   
4   0 -363.823732  2.951346 -3.726368  330.527539  0.010074 -0.000004   

            x11       x12       x13  x14       x15       x16       x17  \
0 -8.900000e-06 -0.806753 -0.805752  401  0.043217  0.001868  0.012497   
1  8.830000e-07 -0.940358 -0.937867  401  0.080432  0.006469 -0.840971   
2 -9.220000e-05  0.240181  0.229315  401  0.274750  0.075487 -0.062232   
3  2.990000e-05 -0.978424 -0.977094  401  0.012307  0.000151  0.227388   
4 -2.060000e-05 -0.915454 -0.907291  401  0.032879  0.001081  0.746032   

        x18        x19  x20  x21  x22  x23       x24       x25       x26  \
0 -0.463844   5.481409   67   78  193  208  0.551122  0.548628  0.127182   
1  5.311628  12.116713   45   65  194  207  0.384040  0.381546  0.376559   
2 -0.579688  14.667095   52   48  204  197  0.411471  0.408978  0.324190   
3 -0.048079   3.178819   36   35  185  216  0.027431  0.024938  0.379052   
4  0.262103   4.029558   94  106  176  225  0.082294  0.079800  0.503741   

        x27       x28       x29       x30       x31       x32       x33  \
0  0.124688  0.002506  0.007481 -0.746494 -2.239481  0.995012  2.172739   
1  0.374065  0.002500  0.004988 -0.883335 -1.766671  0.997506  1.803625   
2  0.321696  0.000000  0.000000  0.000000  0.000000  1.000000  2.121199   
3  0.376559  0.012626  0.024938 -4.876954 -9.753908  0.987531  2.185079   
4  0.501247  0.000000  0.000000  0.000000  0.000000  1.000000  2.063019   

        x34       x35       x36       x37       x38  x39  x40  x41  x42  x43  \
0 -0.689267 -0.906939  0.000347  0.000291  0.000559    1    1    1    1    0   
1 -0.642837 -1.362523  0.005266  0.004333 -0.003138    1    1    1    0    0   
2  0.818939 -0.345764 -0.001580 -0.003453 -0.005586    1    1    1    1    1   
3 -0.936042 -1.012427 -0.000084 -0.000405 -0.000674    1    1    1    1    0   
4 -0.774391 -0.971482  0.000072  0.000002  0.000007    1    1    1    1    0   

   x44  x45       x49       x50       x51       x52       x53       x54  \
0    0    0 -0.863371 -0.844213 -0.825756 -0.815687 -0.797841 -0.783021   
1    0    0 -1.010609 -0.973787 -0.961984 -0.951379 -0.923490 -0.901670   
2    0    0 -0.164767 -0.032252  0.100638  0.198381  0.280628  0.349402   
3    0    0 -0.992626 -0.986964 -0.983844 -0.981017 -0.974313 -0.970775   
4    0    0 -0.943286 -0.935992 -0.930382 -0.924657 -0.902470 -0.888912   

        x55       x56  x57       x58       x59       x60       x61       x62  \
0 -0.766101 -0.746494    1  0.897371  0.839176  0.811325  0.774266  0.730998   
1 -0.886030 -0.861417    1  0.769634  0.307352 -0.039063 -0.120251 -0.014816   
2  0.476402  0.610916    1  0.983643  0.956236  0.930309  0.902098  0.869844   
3 -0.966204 -0.960642    1  0.639823  0.457482  0.394356  0.415160  0.447908   
4 -0.878450 -0.863476    1  0.914871  0.839502  0.779439  0.715020  0.649401   

        x63       x64       x65       x66  x67  x68  x69  x70  x71  x72  x73  \
0  0.704487  0.676231  0.649431  0.648554   28   33  109   42   30    1    1   
1  0.111915  0.139059  0.059254 -0.061575   28   33   80   47   31    1    1   
2  0.835157  0.799799  0.762408  0.723252   30   31   82   24   13    1    1   
3  0.549856  0.525067  0.484631  0.388960   36   40  109   52   35    1    1   
4  0.600535  0.554517  0.520204  0.502750   35   41  110   44   31    1    1   

   x74       x75       x76       x77       x78       x79       x80       x82  \
0    1  2.079063  0.009790  0.012405  0.012023  0.013513  0.013704  0.010211   
1    1  1.407760  0.047294  0.030238  0.024851  0.025847  0.030292  0.006389   
2    1  2.193741  0.026421  0.030931  0.028902  0.030674  0.036668  0.032148 

In [94]:
df_test_dropped.head()

x4          x5        x6         x7          x8        x9       x10  \
0   1    4.665949  0.083412 -17.397081    0.067865  0.005895  0.000012   
1   0 -398.299487  0.784032 -10.785768  403.784446  0.070405  0.000180   
2   1 -380.902140  0.671324  -9.265829  361.841929  0.006937  0.000025   
3   1 -342.784124  0.469414  -8.044287  293.043601  0.007217 -0.000021   
4   1    6.248502  0.252336  -5.398056    0.110097  0.005573  0.000010   

            x11       x12       x13  x14       x15       x16       x17  \
0 -2.090000e-05  0.011444  0.011636  401  0.005818  0.000034  0.034087   
1 -5.890000e-05 -0.977888 -0.993266  401  0.142714  0.020367 -0.319843   
2 -5.380000e-06 -0.949965 -0.949881  401  0.008701  0.000076  0.337449   
3  9.040000e-06 -0.854002 -0.854823  401  0.007698  0.000059 -0.257195   
4  5.300000e-07  0.015457  0.015582  401  0.005635  0.000032  0.047862   

        x18        x19  x20  x21  x22  x23       x24       x25       x26  \
0  0.327009   2.358097    9   13  197  204  0.324190  0.321696  0.553616   
1 -0.281809  28.161913   11   11  215  186  0.927681  0.925187  0.663342   
2  0.615012   2.774954   12   12  198  203  0.099751  0.097257  0.623441   
3 -0.010352   2.886832   10   15  214  187  0.291771  0.289277  0.289277   
4 -0.299650   2.229109    8    8  197  204  0.835411  0.832918  0.688279   

        x27       x28       x29        x30        x31       x32       x33  \
0  0.551122  0.012723  0.032419   0.068020   0.175635  0.980050  2.154294   
1  0.660848  0.000000  0.000000   0.000000   0.000000  1.000000  2.204332   
2  0.620948  0.005025  0.012469  -1.897143  -4.749610  0.992519  2.163365   
3  0.286783  0.058511  0.117207 -18.785227 -40.126251  0.937656  2.177239   
4  0.685786  0.007595  0.022444   0.036548   0.106599  0.985037  2.221545   

        x34       x35           x36           x37           x38  x39  x40  \
0  0.030290 -0.007108 -1.860000e-07 -4.480000e-08  8.590000e-08    1    1   
1 -0.645101 -1.454337  1.375799e-02  1.996644e-02  1.846862e-02    1    1   
2 -0.914883 -0.975463  4.380000e-05  9.340000e-05  8.180000e-05    1    1   
3 -0.831585 -0.875648 -5.750000e-05 -1.061480e-04 -6.650000e-05    1    1   
4  0.031521 -0.001468 -2.930000e-07  1.530000e-07 -3.230000e-08    1    1   

   x41  x42  x43  x44  x45       x49       x50       x51       x52       x53  \
0    1    1    0    0    0  0.003968  0.007107  0.009134  0.010318  0.012917   
1    1    1    0    0    0 -1.185799 -1.123612 -1.065856 -1.013341 -0.948284   
2    1    0    0    0    0 -0.960480 -0.957066 -0.954753 -0.952626 -0.947893   
3    1    1    0    0    0 -0.865476 -0.861306 -0.858338 -0.856174 -0.852460   
4    1    1    0    0    0  0.008355  0.010875  0.012501  0.014046  0.017007   

        x54       x55       x56  x57       x58       x59       x60       x61  \
0  0.014480  0.016416  0.019016    1  0.138492 -0.052568 -0.056833 -0.067865   
1 -0.912662 -0.866746 -0.805297    1  0.785036  0.469576  0.188443 -0.002875   
2 -0.945974 -0.943289 -0.939138    1  0.467677  0.298143  0.180085  0.063116   
3 -0.850653 -0.848182 -0.845795    1  0.235318  0.094647  0.260005  0.073091   
4  0.018515  0.020571  0.023094    1  0.232341 -0.090812  0.021111  0.039492   

        x62       x63       x64       x65       x66  x67  x68  x69  x70  x71  \
0 -0.009167  0.059678  0.014958 -0.023078  0.011155   40   37  121   56   37   
1 -0.161668 -0.283899 -0.397528 -0.471080 -0.492632   23   32   65   39   28   
2  0.033894 -0.035738 -0.070120  0.010165  0.036840   34   38  116   48   30   
3  0.044456  0.098885  0.103982  0.099490  0.076674   39   32  125   54   37   
4  0.055979 -0.000769  0.079791 -0.023135  0.064000   40   39  112   53   33   

   x72  x73  x74       x75       x76       x77       x78       x79       x80  \
0    1    1    1  1.851898  0.003125  0.003654  0.004036  0.004580  0.005895   
1    1    1    1  1.982182  0.072363  0.070047  0.073761  0.071127  0.070405   
2    1    1    1  1.765693  0.003882  0.004508 

In [95]:
df_train_dropped.to_csv('/content/train_data_processed.csv', index=False)
df_test_dropped.to_csv('/content/test_data_processed.csv', index=False)

Quick Note: The saved dataframes are processed dataframes with the sparse columns removed. We have not done any feature scaling or normalization. We dropped the subjects in the test set that are not in the train set.

In [96]:
df_train_dropped.shape, df_test_dropped.shape

((4584, 619), (1732, 618))